### PR TITLE
fix(standard-reports): replace report-table select with async search input

### PR DESCRIPTION
**Pull request by HendrikThePendric**
ToDo before taken out of draft:
- [x] Fix tests
- [x] Resolve issue with POST/PUT either in the app, or on the server (see [Slack conversation](https://dhis2.slack.com/archives/CPGUFVC56/p1614702367109000))

I’ve had some bad experiences in the past with PUT requests clearing fields that are not in the payload. It seems that by adding `mergeMode=MERGE` to the url this can be avoided. I’m going to do some more testing on this locally, but would very much appreciate it if the reviewer could as well.

### CODE REVIEW

**Comment by HendrikThePendric** [comment]
OK, after doing a little research locally, I have found that f9d5505 is not a good solution. I saved the response payload of the `GET` when loading the form, then updated some fields via a `PUT` with `mergeMode=MERGE`, then accessed the same report form again and saved the response payload of the `GET` (again). I then ran a diff between the 2 json files and I noticed that quite a few fields that should have been left untouched actually changed. Here's a summary of the fields that have been updated when they shouldn't have been:
- `created`: I guess this isn't a big deal, just not correct
- `publicAccess`: this changed from `--------` to `rw------`
- `externalAccess`: changed from `true` to `false`
- `userGroupAccesses`: this array was populated before but is empty after the `PUT`

@larshelge could you please advise how to proceed, since `PATCH` threw a 500 error and `PUT` has some issues of its own?

I'm moving this PR back into "Draft" status since I don't think we can merge this in as it stands.

Since I've saved these payloads locally I might as well attach them:
[before_and_after_PUT.zip](https://github.com/dhis2/reports-app/files/6077449/before_and_after_PUT.zip)

File: src/pages/StandardReport.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ class StandardReport extends React.Component {
 
     componentDidMount() {
         if (this.props.reports.length === 0) {
-            this.props.loadStandardReports(true)
+            this.props.loadStandardReports()
```
Comment (ismay) [inline src/pages/StandardReport.js:36]: Ok, so I'm assuming that this was the culprit eventually? As it seems that if this is `true` (in the old code), that for versions above minor 34 will trigger a fetch to the `visualization` endpoint (called `visualizations` in the bug report):

https://github.com/dhis2/reports-app/blob/master/src/utils/backwardCompatability.js#L44-L45

**Comment by larshelge** [comment]
See changes to sharing model here:

https://github.com/dhis2/notes-backend/blob/master/platform/36/changelog/sharing_model.md

File: src/components/form/ReportTableSearchInput.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    const onBlur = () => {
+        if (isSearchMode) {
+            // HACK: when clicking on a list item the onBlur event
+            // of the input happens first and the onClick event on the
+            // list item second. This timeout works around that problem
+            // but we probably need a more structural fix
+            setTimeout(() => {
+                setSearchMode(false)
+            }, 250)
```
Comment (ismay) [inline src/components/form/ReportTableSearchInput.js:85]: Yeah I agree, this seems fragile.

File: src/redux/reducers/standardReportTables.js [hunk 3]
```diff
@@ -2,31 +2,53 @@ import { actionTypes } from '../actions/standardReportTables'
...
+    noMatches: false,
+    searchTerm: '',
 }
 
 export const standardReportTables = (
     state = defaultState,
     { type, payload } = {}
 ) => {
     switch (type) {
+        case actionTypes.STANDARD_REPORT_TABLES_SET_SEARCH_TERM:
```
Comment (ismay) [inline src/redux/reducers/standardReportTables.js:16]: This is more a remark on the existing code style, but referring to the action types in this manner misses on one of the intended benefits of sharing action types as constants: being notified when you're using an action type that doesn't exist. You'll commonly see action types being imported and used directly (not as a collection), because that way you'll get an error if you use a type that doesn't exist.

https://stackoverflow.com/a/37054518

Comment (ismay) [reply src/redux/reducers/standardReportTables.js:16]: But yeah, not really part of the current PR I assume, and probably already on the radar.

**Review by ismay — COMMENTED** [review]
Fix and the ui works when tested with the debug backend. There's no initial unfiltered request to visualizations, only after searching a filtered request is sent.

I do find the existing code style in this app hard to follow. Fair amount of digging required to understand what in the end are fairly straightforward bits of functionality. Would be good to get that standardized a bit because otherwise I can imagine more things like this popping up in the future.

I understand that submitting still has to be fixed so I didn't test that.

File: src/redux/reducers/standardReportTables.js (hunk 3 above)
Comment (HendrikThePendric) [reply src/redux/reducers/standardReportTables.js:16]: 👍

File: src/pages/StandardReport.js (hunk 1 above)
Comment (HendrikThePendric) [reply src/pages/StandardReport.js:36]: Yeah that was one of the culprits, the were actually two, see  57023f2 and 4e32382

Comment (ismay) [reply src/pages/StandardReport.js:36]: Ah ok, I see, thanks 👍

**Review by Mohammer5 — APPROVED** [review]
Looks okay from my end and seems to be working

**Comment by HendrikThePendric** [comment]
I've tested against a v33 backend as well, and this works fine too.

File: src/components/form/ReportTableSearchInput.js (hunk 2 above)
Comment (HendrikThePendric) [reply src/components/form/ReportTableSearchInput.js]: `onMouseDown` happens before the inputs `onBlur`, so I switched to that in https://github.com/dhis2/reports-app/pull/249/commits/37e672bc25c7d7adceee8928f3cac2e93a09bb9e
